src/main/java/ca/nines/ise/writer/XMLWriter.java

```diff
@@ -16,179 +16,1303 @@
+			/* end empty lines */
+			if (normalizeBoundary(lines[i]).isEmpty()){
+				continue;
+			}
+			/* add each line as its own line */
+			xmlStack.ensure_in_line();
+			xmlStack.append_to_line(lines[i]);
+			/* don't end the last line*/
+			if (i < (lines.length - 1))
+				xmlStack.end_line();
+		}
+	}
+
+	/**
+	 * wrapper for creating a HashMap
+	 * 
+	 * @return new HashMap<String,String>
+	 */
+	private HashMap<String, String> new_map() {
+		return new HashMap<String, String>();
+	}
+
+	/**
+	 * Wrapper for putting 2 strings into a HashMap<String,String>
+	 * 
+	 * @param hm
+	 *            HashMap
+	 * @param s1
+	 *            first string
+	 * @param s2
+	 *            second string
+	 * @return HashMap (@hm)
+	 */
+	private HashMap<String, String> map_put(HashMap<String, String> hm,
+			String s1, String s2) {
+		hm.put(s1, s2);
+		return hm;
+	}
+
+	/**
+	 * Sets the attributes of @node to @e
+	 * 
+	 * @param node
+	 *            node with source attributes
+	 * @param e
+	 *            destination element
+	 * @return Element (@e)
+	 */
+	private Element set_attributes(TagNode node, Element e) {
+		return set_attributes(node, e, null, null, null);
+	}
+
+	/**
+	 * Applies attributes to a given Element @e All attributes in @node are
+	 * added to @e with the following caveats: Attributes in @node with a name
+	 * that is a key to @changes have their names changed to the name in @changes
+	 * corresponding to that name key Attributes in @node with a name in @ignore
+	 * are ignored entirely All attributes in @add are also added to @e
+	 * 
+	 * @param node
+	 *            node to copy attributes from
+	 * @param e
+	 *            destination Element to which attributes are added
+	 * @param changes
+	 *            HashMap of attributes whose values are to be changed
+	 * @param ignore
+	 *            Attribute names to be ignored
+	 * @param add
+	 *            extra Attributes to be added to @e
+	 * @return Element (@e)
+	 */
+	private Element set_attributes(TagNode node, Element e,
+			HashMap<String, String> changes, String[] ignore, Attribute[] add) {
+		for (String name : node.getAttributeNames()) {
+			if (ignore != null && Arrays.asList(ignore).contains(name))
+				continue;
+			if (changes != null && changes.containsKey(name))
+				e.addAttribute(new Attribute(changes.get(name), node
+						.getAttribute(name).replace(',', ' ')));
+			else
+				e.addAttribute(new Attribute(name, node.getAttribute(name)
+						.replace(',', ' ')));
+		}
+		if (add != null)
+			for (Attribute attr : add)
+				e.addAttribute(attr);
+		return e;
+	}
+
+	/**
+	 * removes bounding white space (strings which are entirely whitespace)
+	 * 
+	 * @param data
+	 * @return " " if boundary, @data otherwise
+	 */
+	private static String normalizeBoundary(String data) {
+		String temp = normalizeSpace(data);
+		if (temp.equals(" ") || temp.equals(""))
+			return temp;
+		return data;
+	}
+
+	private static String normalizeSpace(String data) {
```

Comment:
Looks like this replaces 2 or more whitespace characters with a single space, is that correct? If it is, then a single regex-replace would be better here: data.replaceAll("\p{Space}\p{Space}+", " ")